syntax = "proto3";
package MasterServer;
message Reply{
  string msg = 1;
  int32 code = 2;
}
message Chunk{
  string id = 1;
  string ip = 2;
  string port = 3;
  int32 num_keys = 4;
  string msg = 5;
  int32 code = 6;
}
message Key{
  string key = 1;
  string chunk = 2;
}
message Replace{
  Chunk old_chunk = 1;
  Chunk new_chunk = 2;
}
message Secondary{
  string id = 1;
  string ip = 2;
  string port = 3;
}
message Directory{
  string key = 1;
  string chunk = 2;
  string msg = 3;
  int32 code = 4;
}
message Primary{
  string ip = 1;
  string port = 2;
}
message Empty{}
service MasterServer{
  rpc heart(Empty) returns (Reply){}
  rpc add_chunk(Chunk) returns (Reply){}
  rpc get_chunk(Key) returns (Chunk){}
  rpc insert_key(Key) returns (Chunk){}
  rpc delete_key(Key) returns (Chunk){}
  rpc replace_chunk(Replace) returns (Reply){}
  rpc add_secondary(Secondary) returns (Reply){}
  rpc sync_directory(Empty) returns (stream Directory){}
  rpc sync_chunks(Empty) returns (stream Chunk){}
  rpc get_primary(Empty) returns (Primary){}
}